### ISSUE TITLE
The value of _valueSetAfterInitialized in SyncVar.cs is always false. 

### ISSUE THREAD

**Issue by KeithSwanger** [FirstGearGames/FishNet#683]
**General**
Fish-Networking version: 4.3.3R

**Description**
In `SyncVar.cs`, `_valueSetAfterInitialized` will never get a chance to be set to `true`

This manifested itself in my project as late-joining clients never getting SyncVars of previously spawned objects synced when they joined.

**Replication**
Take a look at these two methods in SyncVar.cs

```cs
public void SetInitialValues(T value)
{
    _initialValue = value;
    UpdateValues(value);

    if (base.IsInitialized)
        _valueSetAfterInitialized = true;
}
```

```cs
internal void SetValue(T nextValue, bool calledByUser, bool sendRpc = false)
{
...
    if (!base.IsInitialized)
    {
        SetInitialValues(nextValue);
        return;
    }
...
}
```

Looking at these two methods, you can see that `_valueSetAfterInitialized` will never get a chance to be set to `true`, as `SetInitialValues` is only ever called when `base.IsInitialized` is false.

This means that `SyncVar.WriteFull()` will never sync `SyncVars` of non-value types.

**Potential Fix**
Perhaps move the assignment into the `SetValue` method instead:

```cs
public void SetInitialValues(T value)
{
    _initialValue = value;
    UpdateValues(value);
}
```

```cs
internal void SetValue(T nextValue, bool calledByUser, bool sendRpc = false)
{
...
    if (!base.IsInitialized)
    {
        SetInitialValues(nextValue);
        return;
    }
    else
    {
	_valueSetAfterInitialized = true;
    }
...
}
```

**Comment by ooonush** [comment]
I also got a sync error after this change. When the server changes the value before the object is spawned, clients do not receive this change.

**Comment by ooonush** [comment]
If the server changes the initial value before initialization, WriteFull does return on line 351

![image](https://github.com/FirstGearGames/FishNet/assets/72870405/2a516465-7bab-483f-bb1a-c68de749964e)

**Comment by SnaiperoG3D** [comment]
God damn, i have same issue i guess, i just updated to 4.3.3 before release my last game version, and got late joiners bug

**Comment by FirstGearGames** [comment]
> If the server changes the initial value before initialization, WriteFull does return on line 351. Isn't there some way to write Unit Tests so that bugs like this don't occur? It's just not the first time I've had a game crash in my project after a FishNet update :(
> 
> ![image](https://private-user-images.githubusercontent.com/72870405/335055046-2a516465-7bab-483f-bb1a-c68de749964e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcxODg1NzcsIm5iZiI6MTcxNzE4ODI3NywicGF0aCI6Ii83Mjg3MDQwNS8zMzUwNTUwNDYtMmE1MTY0NjUtN2JhYi00ODNmLWJiMWEtYzY4ZGU3NDk5NjRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMxVDIwNDQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4N2Q2ZjE3ZWJlZDY0NDkzMWRiN2U2ZmUwOGM5YWRmNjIzOWY4NWYxNDAzMGMzN2VjM2NjMmFhMzdiMjRmNTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gSoiIBWN8XKlr6cHFtvmmkuNVRLDbpCKQKd8fyLJeYo)



> If the server changes the initial value before initialization, WriteFull does return on line 351. Isn't there some way to write Unit Tests so that bugs like this don't occur? It's just not the first time I've had a game crash in my project after a FishNet update :(
> 
> ![image](https://private-user-images.githubusercontent.com/72870405/335055046-2a516465-7bab-483f-bb1a-c68de749964e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcxODg1NzcsIm5iZiI6MTcxNzE4ODI3NywicGF0aCI6Ii83Mjg3MDQwNS8zMzUwNTUwNDYtMmE1MTY0NjUtN2JhYi00ODNmLWJiMWEtYzY4ZGU3NDk5NjRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMxVDIwNDQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4N2Q2ZjE3ZWJlZDY0NDkzMWRiN2U2ZmUwOGM5YWRmNjIzOWY4NWYxNDAzMGMzN2VjM2NjMmFhMzdiMjRmNTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gSoiIBWN8XKlr6cHFtvmmkuNVRLDbpCKQKd8fyLJeYo)



> If the server changes the initial value before initialization, WriteFull does return on line 351. Isn't there some way to write Unit Tests so that bugs like this don't occur? It's just not the first time I've had a game crash in my project after a FishNet update :(
> 
> ![image](https://private-user-images.githubusercontent.com/72870405/335055046-2a516465-7bab-483f-bb1a-c68de749964e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcxODg1NzcsIm5iZiI6MTcxNzE4ODI3NywicGF0aCI6Ii83Mjg3MDQwNS8zMzUwNTUwNDYtMmE1MTY0NjUtN2JhYi00ODNmLWJiMWEtYzY4ZGU3NDk5NjRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMxVDIwNDQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4N2Q2ZjE3ZWJlZDY0NDkzMWRiN2U2ZmUwOGM5YWRmNjIzOWY4NWYxNDAzMGMzN2VjM2NjMmFhMzdiMjRmNTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gSoiIBWN8XKlr6cHFtvmmkuNVRLDbpCKQKd8fyLJeYo)

It's near impossible to provide unit tests for scenarios we don't know could cause problems. When we made the changes they were tested and all tests passed. If you can let me know the condition which causes this as code, I can include it in the existing tests.

**Comment by FirstGearGames** [comment]
I'm going to get this resolved and released by tomorrow. I'm going to also open up bounties for help on creating tests for situations which cannot simply be done through normal unit tests.

**Comment by FirstGearGames** [comment]
Thank you @KeithSwanger for finding this issue as well the solution. 
Setting to true inside SetValue should work without issue. This is resolved in 4.3.4.
